core/src/main/java/io/aiven/kafka/tieredstorage/chunkmanager/CachingChunkManager.java

```diff
@@ -0,0 +1,133 @@
+import org.slf4j.LoggerFactory;
+
+public abstract class CachingChunkManager<T> implements Configurable, ChunkManager {
+    private static final Logger log = LoggerFactory.getLogger(CachingChunkManager.class);
+    private static final long GET_TIMEOUT_SEC = 10;
+    private final ChunkManager chunkManager;
+    private final Executor executor;
+    protected AsyncCache<ChunkKey, T> cache;
+
+    public CachingChunkManager(final ChunkManager chunkManager, final Executor executor) {
+        this.chunkManager = chunkManager;
+        this.executor = executor;
+    }
+
+    /**
+     * Fetches a specific chunk from remote storage and stores into the cache.
+     * Since it's not possible to cache an opened InputStream the actual data is cached and everytime
+     * there is a call to cache the InputStream is recreated from the data stored in cache and stored into local
+     * variable. This also allows to solve the race condition between eviction and fetching. Since the InputStream is
+     * opened right when fetching from cache happens even if the actual value will be removed from the cache
+     * the InputStream will still contain the data.
+     */
+    public InputStream getChunk(final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+                                final SegmentManifest manifest,
+                                final int chunkId) throws StorageBackendException, IOException {
+        final Uuid id = remoteLogSegmentMetadata.remoteLogSegmentId().id();
+        final ChunkKey chunkKey = new ChunkKey(id, chunkId);
+        final AtomicReference<InputStream> result = new AtomicReference<>();
+        try {
+            return cache.asMap().compute(chunkKey, (key, val) -> CompletableFuture.supplyAsync(() -> {
+                if (val == null) {
+                    try {
+                        final InputStream chunk = chunkManager.getChunk(remoteLogSegmentMetadata, manifest, chunkId);
+                        final T t = this.cacheChunk(chunkKey, chunk);
+                        result.getAndSet(chunkToStream(t));
+                        return t;
+                    } catch (final StorageBackendException | IOException e) {
+                        throw new CompletionException(e);
+                    }
+                } else {
+                    try {
+                        final T cachedChunk = val.get();
+                        result.getAndSet(chunkToStream(cachedChunk));
+                        return cachedChunk;
+                    } catch (InterruptedException | ExecutionException e) {
+                        throw new CompletionException(e);
+                    }
+                }
+            }, executor)).thenApplyAsync(t -> {
+                log.info("Returning stream");
+                return result.get();
+            }).get(GET_TIMEOUT_SEC, TimeUnit.SECONDS);
+        } catch (final ExecutionException e) {
+            // Unwrap previously wrapped exceptions if possible.
+
```

Comment:
nit: rm line?

Reply:
Keeping the original style since this is a copy+paste, will address formatting in both places together in a separate PR.